### DISCUSSION TITLE
[General] How to Retrieve Proof Parameters for ETH Claim on L1 from zkSync Era?

### DISCUSSION THREAD

**Discussion by mihailo-maksa** [zkSync-Community-Hub/zksync-developers#812] [General] [answered]
### Team or Project

Kelp DAO

### Environment

Mainnet

### Provide a brief description of the functionality you're trying to implement and the issue you are running into.

We're implementing a functionality in our contracts to bridge ETH from zkSync Era (L2) to Ethereum Mainnet (L1). However, we're encountering challenges with the claim transaction on Mainnet. Specifically, we're unsure how to obtain the correct proof parameters required for the L1 `finalizeEthWithdrawal` function, which must be called after the 24-hour bridging delay. Is there an API or other recommended and reliable method to fetch these proof parameters?

**Comment by alhenaai** [comment]
To retrieve the proof parameters required for the `finalizeEthWithdrawal` function on Ethereum Mainnet after bridging ETH from ZKsync Era, you can utilize the Prover API provided by ZKsync. This API is designed to help you fetch batch proof inputs that are necessary for finalizing withdrawals on L1.

Here are the relevant endpoints you might find useful:

1. **/proof_generation_data** - This endpoint allows you to retrieve the latest batch proof inputs, which are very similar to the internal workings of ZKsync's provers.
2. **/proof_generation_data/{batch_number}** - Use this endpoint to get inputs for a specific batch proof, which is useful for testing or benchmarking purposes.

Please note that not all batches have proof inputs available, and batch data is only available for a finite amount of time (30 days from the time of writing). After this period, the data is deleted, and ZKsync cannot verify batches older than this timeframe.

For more detailed information on setting up and using the Prover API, you can refer to the [From Integration to Verification: Completing the First Steps in ZKsync](https://zksync.mirror.xyz/1HzbsDVMQeE3P-DgysjKzagNnHehwCIrC1QA9XcNRnk) documentation.

For further discussions and support, you may also visit the [ZKsync developers discussions on Github](https://github.com/zkSync-Community-Hub/zksync-developers/discussions).

Your feedback helps us serve you better. If this fully addresses your question, please give it a thumbs up. If it doesn't, please give it a thumbs down. This will inform the community and team so we can improve our support.

**Comment by bxpana** [comment]
Thanks for the post! Will check this out and get back to you

**Reply by mihailo-maksa** [reply]
Sounds good, thank you!

**Comment by popzxc** [comment, marked as answer]
Hey @mihailo-maksa! Yes, there are APIs for that.
You can take a look at how fetching withdrawal parameters is [implemented in zksync-ethers](https://github.com/zksync-sdk/zksync-ethers/blob/main/src/adapters.ts#L1448). Corresponding RPC method is[zks_getL2ToL1LogProof](https://docs.zksync.io/zksync-protocol/api/zks-rpc#zks_getl2tol1logproof).

Basically, what you need is to identify the log that corresponds to the withdrawal L2 -> L1 message _on l2_ and then retreive a proof for it.

One caveat you should be aware of is that there might be multiple withdrawals in a single transaction (which is why `index` needs to be passed in the function above).

**Reply by mihailo-maksa** [reply]
Hey @popzxc, thanks for sharing this, it was very helpful!

I was also wondering if this would work in case in which a call to `withdraw` on L2 is done via some other function on the L2 smart contract? 

e.g. suppose that there is an `L2Vault` contract on zkSync Era that has a `bridgeAssets` method on it, callable only by the contract `owner`, and within this `bridgeAssets` function, there is a call to `withdraw` to initiate the withdrawal of ETH from L2 --> L1

**Reply by popzxc** [reply]
Absolutely, you only need to extract the log that corresponds to the L2->L1 message initiating a withdrawal. How withdrawal was initiated is not important.
Moreover, the algorithm is the same for any form of L2->L1 communication.

**Reply by mihailo-maksa** [reply]
Sounds good, thank you for the detailed answer!
